File: leetcode/medium/46. Permutations/solution.cc
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        pick(nums, 0, res);
        return res;
    }
private:
    void pick(vector<int>& nums, int depth, vector<vector<int>>& res) {
        if (depth >= nums.size()) {
            res.push_back(nums);
            return;
        }

        for (int i = depth; i < nums.size(); i += 1) {
            swap(nums[depth], nums[i]);
            pick(nums, depth + 1, res);
            swap(nums[depth], nums[i]);
        }
    }
};

File: leetcode/medium/46. Permutations/solution.js
// https://leetcode.com/problems/permutations/
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
const permute = (nums) => {
    const permuts = [];
    const isVisit = Array(nums.length).fill(false);
    let temp = [];

    const pick = (depth) => {
        if (temp.length === depth) {
            permuts.push(Array.from(temp));
            return;
        }

        for (let i = 0; i < depth; i += 1) {
            if (!isVisit[i]) {
                isVisit[i] = true;
                temp.push(nums[i]);
                pick(depth);
                temp.pop();
                isVisit[i] = false;
            }
        }
    };

    pick(nums.length);
    return permuts;
};

permute([1,2,3])